comp/core/flare/profiling.go

```diff
@@ -0,0 +1,76 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package flare
+
+import (
+	"fmt"
+
+	"github.com/fatih/color"
+
+	"github.com/DataDog/datadog-agent/pkg/api/util"
+	"github.com/DataDog/datadog-agent/pkg/config/settings"
+)
+
+// ProfilingOpts defines the options used for profiling
+type ProfilingOpts struct {
+	ProfileMutex         bool
+	ProfileMutexFraction int
+	ProfileBlocking      bool
+	ProfileBlockingRate  int
+}
+
+// SetRuntimeProfilingSettings sets the runtime settings used for profiling
+func SetRuntimeProfilingSettings(opts ProfilingOpts, settingsClient settings.Client) (func(), error) {
+	if err := util.SetAuthToken(); err != nil {
+		return nil, fmt.Errorf("unable to set up authentication token: %v", err)
+	}
+
+	prev := make(map[string]interface{})
+	if opts.ProfileMutex && opts.ProfileMutexFraction > 0 {
+		old, err := setRuntimeSetting(settingsClient, "runtime_mutex_profile_fraction", opts.ProfileMutexFraction)
+		if err != nil {
+			return nil, err
+		}
+		prev["runtime_mutex_profile_fraction"] = old
+	}
+	if opts.ProfileBlocking && opts.ProfileBlockingRate > 0 {
+		old, err := setRuntimeSetting(settingsClient, "runtime_block_profile_rate", opts.ProfileBlockingRate)
+		if err != nil {
+			return nil, err
+		}
+		prev["runtime_block_profile_rate"] = old
+	}
+
+	return func() { resetRuntimeProfilingSettings(prev, settingsClient) }, nil
+}
+
+func setRuntimeSetting(c settings.Client, name string, value int) (interface{}, error) {
+	fmt.Fprintln(color.Output, color.BlueString("Setting %s to %v", name, value))
+
+	oldVal, err := c.Get(name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current value of %s: %v", name, err)
+	}
+
+	if _, err := c.Set(name, fmt.Sprint(value)); err != nil {
+		return nil, fmt.Errorf("failed to set %s to %v: %v", name, value, err)
+	}
+
+	return oldVal, nil
+}
+
+func resetRuntimeProfilingSettings(prev map[string]interface{}, settingsClient settings.Client) {
```

Comment:
Could we maybe have a single function so set/reset settings ?

Something like `ExecWIthSetting(callback, setting...)` that would set the setting, exec the callback and reset everything back afterward ?

WDYT ?